src/coffee/sync/product-sync.coffee

```diff
@@ -37,12 +37,20 @@ class ProductSync extends BaseSync
     super new_obj, old_obj
 
   _doMapActions: (diff, new_obj, old_obj) ->
+    # New variant can have attributes with new values and this can cause conflict if the product still
+    # has old value of the sameForAll attribute. That's why it's necessary to update the attributes
+    # before adding new variant
+    variantActions = @_mapActionOrNot 'variants', => @_utils.actionsMapVariants(diff, old_obj, new_obj)
```

Comment:
I think it's a bit more readable. WDYT?

```js
allActions.push variantActions.filter (action) -> action.action is 'removeVariant'
allActions.push @_mapActionOrNot 'attributes', => @_utils.actionsMapAttributes(diff, old_obj, new_obj, @sameForAllAttributeNames)
allActions.push variantActions.filter (action) -> action.action is 'addVariant'
```
vs
```js
allActions.push removeVariantActions
allActions.push @_mapActionOrNot 'attributes', => @_utils.actionsMapAttributes(diff, old_obj, new_obj, @sameForAllAttributeNames)
allActions.push addVariantActions
```

But it's not a problem for me to change it as you suggested.